Add basic support for legacy version files (e.g. .ruby-version)

Stratus3D opened the pull request:
Add basic support for legacy version files (e.g. .ruby-version) via an optional `get_version_from_legacy_file` script in each plugin.

Questions:
* Is it safe to equate plugin name with tool name? Or could there be variation in the tool name? (I make this assumption in the `get_tool_version_from_legacy_file` function).
* Should `get_version_from_legacy_file` be named `get-version-from-legacy-file`? It seems the other "callback" scripts use dashes. I'll be happy to change this. I only noticed this after I finished.

I have manually tested this PR with the updated asdf-ruby plugin on Ubuntu 14.04 and OSX 10.10.5.

Stratus3D commented:
This does not add support for legacy version files to the install command. I wasn't sure if that would be a good idea or not.

HashNuke commented:
@Stratus3D Thank you. I'll check this out and merge once I write tests for this. Need a week to finish up other work. I also have to add a config option to enable/disable this feature.

HashNuke commented:
@Stratus3D Yes please. dasherized names. Sorry about that I forgot I use dashes everywhere I can avoid underscores.

Yes the assumption is plugin name is tool name. I think if you add the ruby plugin repo as "foobar" plugin, then your .tool-version files will be "foobar 2.1.2" instead of "ruby 2.1.2". Hence the convention for names of repositories. "asdf-foobar" github repo name means the user is assumed to add it as a foobar plugin.

Stratus3D commented:
@HashNuke I updated the PR so it expects the legacy version script to use dashes. I also renamed the file in the asdf-ruby plugin: https://github.com/HashNuke/asdf-ruby/pull/4

Stratus3D commented:
Should I add legacy version support to the install command? I could add it in a separate PR if you wanted.

HashNuke commented:
ah no worries. This amount of detail is enough. We just want to make the
transition/co-operation between other version managers easier

On Sun, Nov 29, 2015 at 10:42 PM, Trevor Brown <notifications@github.com>
wrote:

> Should I add legacy version support to the install command? I could add it
> in a separate PR if you wanted.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HashNuke/asdf/pull/24#issuecomment-160432620>.
>

HashNuke commented:
I promise to finish up my work this week.

TODO:
* [ ] setup tests
* [ ] support config file

Stratus3D commented:
Please do not merge this. I believe I discovered a bug while using my forked version containing these changes. I will update the PR when I figure out the issue.

HashNuke commented:
@Stratus3D No worries. Thanks for updating the issue :)

colinrymer commented:
Any update on this? Anything I can do to help? I'm very interested in using this, but support for `.ruby-version` and `.node-version` files is a prerequisite.

HashNuke commented:
@colinrymer I'll try to get this in this weekend. @Stratus3D needs to confirm too.

colinrymer commented:
I think I see the bug @Stratus3D was talking about (and if not, then identified another): the presence of a `.ruby-version`, etc. file should be checked in parent directories if a `.tool-version` file can't be found higher up as well.

Stratus3D commented:
@colinrymer I believe that's by design. I we decided to only check for the legacy version file in the current directory. If it's not found in the current directory we resort to searching for the `.tool-versions` file in parent directories.

I think the bug I found was due to what happens when a `.ruby-version` file isn't present. It wasn't kicking back over to what's defined in the `.tool-versions` file. Not sure what was causing that though. Hopefully I'll get it figured out this weekend.

HashNuke commented:
@Stratus3D Oh yes I remember that conversation now. Thanks for clarifying Trevor ~!

Stratus3D commented:
I just pushed a change that fixes the issue I was experiencing. @HashNuke @colinrymer do you guys want to test it out on your machines? You'll also need asdf-ruby, since that's the only plugin that uses this functionality right now.

Stratus3D commented:
Made another mistake in the get_tool_version_from_legacy_file function. Just pushed the fix.

HashNuke commented:
@Stratus3D Thanks Trevor. I'll test it out in a few hours.

Stratus3D commented:
@HashNuke let me know if there is anything else I can do to help with this PR. If you need tests written let me know.

HashNuke commented:
@Stratus3D Sorry abt the delay in reply. PR is fine. This needs config file support before it gets merged - https://github.com/asdf-vm/asdf/issues/41

Created an issue and marked it as "help wanted".

tuvistavie commented:
@Stratus3D config file support just go merged, so I think this should be ready to go too!

I think you would just need to add the following condition when getting the version from the legacy file.

`$(get_asdf_config_value "legacy_version_file") = "yes"`

HashNuke commented:
@tuvistavie @Stratus3D Resolved conflicts in PR #47 and added the check for config.

Stratus3D commented:
Thanks @HashNuke !
